In [1]:
import pandas as pd

In [8]:
df = pd.read_csv('STEG_BILLING_HISTORY.csv')
df

C:\Users\Ummhabibi\AppData\Local\Temp\ipykernel_12380\1315909486.py:1: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('STEG_BILLING_HISTORY.csv')


client_id invoice_date  tarif_type  counter_number  \
0            train_Client_0   2014-03-24          11       1335667.0   
1            train_Client_0   2013-03-29          11       1335667.0   
2            train_Client_0   2015-03-23          11       1335667.0   
3            train_Client_0   2015-07-13          11       1335667.0   
4            train_Client_0   2016-11-17          11       1335667.0   
...                     ...          ...         ...             ...   
4476744  train_Client_99998   2005-08-19          10       1253571.0   
4476745  train_Client_99998   2005-12-19          10       1253571.0   
4476746  train_Client_99999   1996-09-25          11        560948.0   
4476747  train_Client_99999   1996-05-28          11        560948.0   
4476748  train_Client_99999   1996-01-25          11        560948.0   

        counter_statue  counter_code  reading_remarque  counter_coefficient  \
0                    0           203               8.0                    1   
1                    0           203               6.0                    1   
2                    0           203               8.0                    1   
3                    0           207               8.0                    1   
4                    0           207               9.0                    1   
...                ...           ...               ...                  ...   
4476744              0           202               9.0                    1   
4476745              0           202               6.0                    1   
4476746              0           203               6.0                    1   
4476747              0           203               6.0                    1   
4476748              0           203               6.0                    1   

         consommation_level_1  consommation_level_2  consommation_level_3  \
0                          82                     0                     0   
1                        1200                   184                     0   
2                         123                     0                     0   
3                         102                     0                     0   
4                         572                     0                     0   
...                       ...                   ...                   ...   
4476744                   400                   135                     0   
4476745                   200                     6                     0   
4476746                   259                     0                     0   
4476747                   603                     0                     0   
4476748                   516                     0                     0   

         consommation_level_4  old_index  new_index  months_number  \
0                           0      14302      14384              4   
1                           0      12294      13678              4   
2                           0      14624      14747              4   
3                           0      14747      14849              4   
4                           0      15066      15638             12   
...                       ...        ...        ...            ...   
4476744                     0       3197       3732              8   
4476745                     0       3732       3938              4   
4476746                     0      13884      14143              4   
4476747                     0      13281      13884              4   
4476748                     0      12765      13281              4   

        counter_type  
0               ELEC  
1               ELEC  
2               ELEC  
3               ELEC  
4               ELEC  
...              ...  
4476744         ELEC  
4476745         ELEC  
4476746         ELEC  
4476747         ELEC  
4476748         ELEC  

[4476749 rows x 16 columns]

In [11]:
#Load the dataset, display the ten first lines, store the results in a variable called 'client_0_bills'
client_0_bills = df.head(10)
client_0_bills

client_id invoice_date  tarif_type  counter_number counter_statue  \
0  train_Client_0   2014-03-24          11       1335667.0              0   
1  train_Client_0   2013-03-29          11       1335667.0              0   
2  train_Client_0   2015-03-23          11       1335667.0              0   
3  train_Client_0   2015-07-13          11       1335667.0              0   
4  train_Client_0   2016-11-17          11       1335667.0              0   
5  train_Client_0   2017-07-17          11       1335667.0              0   
6  train_Client_0   2018-12-07          11       1335667.0              0   
7  train_Client_0   2019-03-19          11       1335667.0              0   
8  train_Client_0   2011-07-22          11       1335667.0              0   
9  train_Client_0   2011-11-22          11       1335667.0              0   

   counter_code  reading_remarque  counter_coefficient  consommation_level_1  \
0           203               8.0                    1                    82   
1           203               6.0                    1                  1200   
2           203               8.0                    1                   123   
3           207               8.0                    1                   102   
4           207               9.0                    1                   572   
5           207               9.0                    1                   314   
6           207               9.0                    1                   541   
7           207               9.0                    1                   585   
8           203               9.0                    1                  1200   
9           203               6.0                    1                  1082   

   consommation_level_2  consommation_level_3  consommation_level_4  \
0                     0                     0                     0   
1                   184                     0                     0   
2                     0                     0                     0   
3                     0                     0                     0   
4                     0                     0                     0   
5                     0                     0                     0   
6                     0                     0                     0   
7                     0                     0                     0   
8                   186                     0                     0   
9                     0                     0                     0   

   old_index  new_index  months_number counter_type  
0      14302      14384              4         ELEC  
1      12294      13678              4         ELEC  
2      14624      14747              4         ELEC  
3      14747      14849              4         ELEC  
4      15066      15638             12         ELEC  
5      15638      15952              8         ELEC  
6      15952      16493             12         ELEC  
7      16493      17078              8         ELEC  
8       7770       9156              4         ELEC  
9       9156      10238              4         ELEC

In [13]:
#What is the data type of the 'client_0_bills' variable ?
client_0_bills.dtypes

client_id                object
invoice_date             object
tarif_type                int64
counter_number          float64
counter_statue           object
counter_code              int64
reading_remarque        float64
counter_coefficient       int64
consommation_level_1      int64
consommation_level_2      int64
consommation_level_3      int64
consommation_level_4      int64
old_index                 int64
new_index                 int64
months_number             int64
counter_type             object
dtype: object

In [14]:
#Display the general information of the dataset and try to answer the following questions : 
#How many rows and columns do we have in this dataset ?
#How many categorical features are present in the dataset ?
#How much memory space does the dataset consume ?

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4476749 entries, 0 to 4476748
Data columns (total 16 columns):
 #   Column                Dtype  
---  ------                -----  
 0   client_id             object 
 1   invoice_date          object 
 2   tarif_type            int64  
 3   counter_number        float64
 4   counter_statue        object 
 5   counter_code          int64  
 6   reading_remarque      float64
 7   counter_coefficient   int64  
 8   consommation_level_1  int64  
 9   consommation_level_2  int64  
 10  consommation_level_3  int64  
 11  consommation_level_4  int64  
 12  old_index             int64  
 13  new_index             int64  
 14  months_number         int64  
 15  counter_type          object 
dtypes: float64(2), int64(10), object(4)
memory usage: 546.5+ MB


How many rows and columns do we have in this dataset ?

16 columns and 4476749 rows

How many categorical features are present in the dataset ?

4

How much memory space does the dataset consume ?

546.5+ MB

In [16]:
#Inspect the dataset for potential missing values.
print(df.isnull().sum())

client_id                  0
invoice_date               0
tarif_type                 0
counter_number            48
counter_statue             0
counter_code               0
reading_remarque        4531
counter_coefficient        0
consommation_level_1       0
consommation_level_2       0
consommation_level_3       0
consommation_level_4       0
old_index                  0
new_index                  0
months_number              0
counter_type               0
dtype: int64


#### Select your strategy to handle missing values, and tell us why you had made that choice.

Two columns have missing values; 'reading_remarque' and 'counter_number'. 

The 'reading_remarque' column has a lot of missing values(4531), ergo filling in the missing values may introduce significant bias. As such, dropping the entire column is the best option.

The 'counter_number' has fewer missing values(48); and since the data type is float, I'd opt to fill in the missing values with the mean value of the column.








In [ ]:
#Dropping 'reading_remarque' column

df.drop(columns=['reading_remarque'], inplace=True)

In [19]:
#Verify column drop
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4476749 entries, 0 to 4476748
Data columns (total 15 columns):
 #   Column                Dtype  
---  ------                -----  
 0   client_id             object 
 1   invoice_date          object 
 2   tarif_type            int64  
 3   counter_number        float64
 4   counter_statue        object 
 5   counter_code          int64  
 6   counter_coefficient   int64  
 7   consommation_level_1  int64  
 8   consommation_level_2  int64  
 9   consommation_level_3  int64  
 10  consommation_level_4  int64  
 11  old_index             int64  
 12  new_index             int64  
 13  months_number         int64  
 14  counter_type          object 
dtypes: float64(1), int64(10), object(4)
memory usage: 512.3+ MB


In [21]:
#Fill missing values in the column with the mean value of the column

df['counter_number'].fillna(df['counter_number'].mean(), inplace=True)

In [23]:
#Verify change
print(df.isnull().sum())

client_id               0
invoice_date            0
tarif_type              0
counter_number          0
counter_statue          0
counter_code            0
counter_coefficient     0
consommation_level_1    0
consommation_level_2    0
consommation_level_3    0
consommation_level_4    0
old_index               0
new_index               0
months_number           0
counter_type            0
dtype: int64


In [26]:
#Run a descriptive analysis on numeric features (columns)
df.describe()

tarif_type  counter_number  counter_code  counter_coefficient  \
count  4.476749e+06    4.476749e+06  4.476749e+06         4.476749e+06   
mean   2.012804e+01    1.230595e+11  1.724884e+02         1.003040e+00   
std    1.347256e+01    1.657267e+12  1.338871e+02         3.083466e-01   
min    8.000000e+00    0.000000e+00  0.000000e+00         0.000000e+00   
25%    1.100000e+01    1.211080e+05  5.000000e+00         1.000000e+00   
50%    1.100000e+01    4.945690e+05  2.030000e+02         1.000000e+00   
75%    4.000000e+01    1.115217e+06  2.070000e+02         1.000000e+00   
max    4.500000e+01    2.798115e+13  6.000000e+02         5.000000e+01   

       consommation_level_1  consommation_level_2  consommation_level_3  \
count          4.476749e+06          4.476749e+06          4.476749e+06   
mean           4.109795e+02          1.093225e+02          2.030620e+01   
std            7.573080e+02          1.220123e+03          1.574239e+02   
min            0.000000e+00          0.000000e+00          0.000000e+00   
25%            7.900000e+01          0.000000e+00          0.000000e+00   
50%            2.740000e+02          0.000000e+00          0.000000e+00   
75%            6.000000e+02          0.000000e+00          0.000000e+00   
max            9.999100e+05          9.990730e+05          6.449200e+04   

       consommation_level_4     old_index     new_index  months_number  
count          4.476749e+06  4.476749e+06  4.476749e+06   4.476749e+06  
mean           5.292588e+01  1.776700e+04  1.834970e+04   4.483095e+01  
std            8.754725e+02  4.036693e+04  4.095321e+04   3.128335e+03  
min            0.000000e+00  0.000000e+00  0.000000e+00   0.000000e+00  
25%            0.000000e+00  1.791000e+03  2.056000e+03   4.000000e+00  
50%            0.000000e+00  7.690000e+03  8.192000e+03   4.000000e+00  
75%            0.000000e+00  2.166000e+04  2.234300e+04   4.000000e+00  
max            5.479460e+05  2.800280e+06  2.870972e+06   6.366240e+05

In [31]:
#Select the bills records for the client with an id 'train_Client_0', using 2 methods.
#Method 1
bill = df[df['client_id'] == 'train_Client_0']
print(bill)

         client_id invoice_date  tarif_type  counter_number counter_statue  \
0   train_Client_0   2014-03-24          11       1335667.0              0   
1   train_Client_0   2013-03-29          11       1335667.0              0   
2   train_Client_0   2015-03-23          11       1335667.0              0   
3   train_Client_0   2015-07-13          11       1335667.0              0   
4   train_Client_0   2016-11-17          11       1335667.0              0   
5   train_Client_0   2017-07-17          11       1335667.0              0   
6   train_Client_0   2018-12-07          11       1335667.0              0   
7   train_Client_0   2019-03-19          11       1335667.0              0   
8   train_Client_0   2011-07-22          11       1335667.0              0   
9   train_Client_0   2011-11-22          11       1335667.0              0   
10  train_Client_0   2008-11-25          11       1335667.0              0   
11  train_Client_0   2012-07-23          11       1335667.0     

In [36]:
#Method 2
df.loc[df['client_id'] == 'train_Client_0']

client_id invoice_date  tarif_type  counter_number counter_statue  \
0   train_Client_0   2014-03-24          11       1335667.0              0   
1   train_Client_0   2013-03-29          11       1335667.0              0   
2   train_Client_0   2015-03-23          11       1335667.0              0   
3   train_Client_0   2015-07-13          11       1335667.0              0   
4   train_Client_0   2016-11-17          11       1335667.0              0   
5   train_Client_0   2017-07-17          11       1335667.0              0   
6   train_Client_0   2018-12-07          11       1335667.0              0   
7   train_Client_0   2019-03-19          11       1335667.0              0   
8   train_Client_0   2011-07-22          11       1335667.0              0   
9   train_Client_0   2011-11-22          11       1335667.0              0   
10  train_Client_0   2008-11-25          11       1335667.0              0   
11  train_Client_0   2012-07-23          11       1335667.0              0   
12  train_Client_0   2012-11-29          11       1335667.0              0   
13  train_Client_0   2013-07-22          11       1335667.0              0   
14  train_Client_0   2014-07-18          11       1335667.0              0   
15  train_Client_0   2014-11-24          11       1335667.0              0   
16  train_Client_0   2015-11-18          11       1335667.0              0   
17  train_Client_0   2012-03-27          11       1335667.0              0   
18  train_Client_0   2011-03-30          11       1335667.0              0   
19  train_Client_0   2013-02-12          11       1335667.0              0   
20  train_Client_0   2010-07-22          11       1335667.0              0   
21  train_Client_0   2010-11-24          11       1335667.0              0   
22  train_Client_0   2005-10-17          11       1335667.0              0   
23  train_Client_0   2006-02-24          11       1335667.0              0   
24  train_Client_0   2006-06-23          11       1335667.0              0   
25  train_Client_0   2006-10-18          11       1335667.0              0   
26  train_Client_0   2007-06-27          11       1335667.0              0   
27  train_Client_0   2007-10-25          11       1335667.0              0   
28  train_Client_0   2007-02-26          11       1335667.0              0   
29  train_Client_0   2008-07-28          11       1335667.0              0   
30  train_Client_0   2009-01-04          11       1335667.0              0   
31  train_Client_0   2009-07-27          11       1335667.0              0   
32  train_Client_0   2009-11-24          11       1335667.0              0   
33  train_Client_0   2010-03-29          11       1335667.0              0   
34  train_Client_0   2008-01-04          11       1335667.0              0   

    counter_code  counter_coefficient  consommation_level_1  \
0            203                    1                    82   
1            203                    1                  1200   
2            203                    1                   123   
3            207                    1                   102   
4            207                    1                   572   
5            207                    1                   314   
6            207                    1                   541   
7            207                    1                   585   
8            203                    1                  1200   
9            203                    1                  1082   
10           203                    1                   174   
11           203                    1                   628   
12           203                    1                   886   
13           203                    1                   147   
14           203                    1                    99   
15           203                    1                   141   
16           207                    1                   217   
17           203                    1                   292   
18 

In [43]:
#Transform the 'counter_type' feature to a numeric variable using the encoder of your choice.
from sklearn.preprocessing import LabelEncoder

In [48]:
le = LabelEncoder()
df['counter_type'] = le.fit_transform(df['counter_type'])
df.head(10).T

0               1               2  \
client_id             train_Client_0  train_Client_0  train_Client_0   
invoice_date              2014-03-24      2013-03-29      2015-03-23   
tarif_type                        11              11              11   
counter_number             1335667.0       1335667.0       1335667.0   
counter_statue                     0               0               0   
counter_code                     203             203             203   
counter_coefficient                1               1               1   
consommation_level_1              82            1200             123   
consommation_level_2               0             184               0   
consommation_level_3               0               0               0   
consommation_level_4               0               0               0   
old_index                      14302           12294           14624   
new_index                      14384           13678           14747   
months_number                      4               4               4   
counter_type                       0               0               0   

                                   3               4               5  \
client_id             train_Client_0  train_Client_0  train_Client_0   
invoice_date              2015-07-13      2016-11-17      2017-07-17   
tarif_type                        11              11              11   
counter_number             1335667.0       1335667.0       1335667.0   
counter_statue                     0               0               0   
counter_code                     207             207             207   
counter_coefficient                1               1               1   
consommation_level_1             102             572             314   
consommation_level_2               0               0               0   
consommation_level_3               0               0               0   
consommation_level_4               0               0               0   
old_index                      14747           15066           15638   
new_index                      14849           15638           15952   
months_number                      4              12               8   
counter_type                       0               0               0   

                                   6               7               8  \
client_id             train_Client_0  train_Client_0  train_Client_0   
invoice_date              2018-12-07      2019-03-19      2011-07-22   
tarif_type                        11              11              11   
counter_number             1335667.0       1335667.0       1335667.0   
counter_statue                     0               0               0   
counter_code                     207             207             203   
counter_coefficient                1               1               1   
consommation_level_1             541             585            1200   
consommation_level_2               0               0             186   
consommation_level_3               0               0               0   
consommation_level_4               0               0               0   
old_index                      15952           16493            7770   
new_index                      16493           17078            9156   
months_number                     12               8               4   
counter_type                       0               0               0   

                                   9  
client_id             train_Client_0  
invoice_date              2011-11-22  
tarif_type                        11  
counter_number             1335667.0  
counter_statue                     0  
counter_code                     203  
counter_coefficient                1  
consommation_level_1            1082  
consommation_level_2               0  
consommation_level_3               0  
consommation_level_4               0  
old_index                       9156  
new_index                      10238  
months_number                 

In [50]:
#Delete the 'counter_statue' feature from the Dataframe
df.drop('counter_statue', axis=1)

client_id invoice_date  tarif_type  counter_number  \
0            train_Client_0   2014-03-24          11       1335667.0   
1            train_Client_0   2013-03-29          11       1335667.0   
2            train_Client_0   2015-03-23          11       1335667.0   
3            train_Client_0   2015-07-13          11       1335667.0   
4            train_Client_0   2016-11-17          11       1335667.0   
...                     ...          ...         ...             ...   
4476744  train_Client_99998   2005-08-19          10       1253571.0   
4476745  train_Client_99998   2005-12-19          10       1253571.0   
4476746  train_Client_99999   1996-09-25          11        560948.0   
4476747  train_Client_99999   1996-05-28          11        560948.0   
4476748  train_Client_99999   1996-01-25          11        560948.0   

         counter_code  counter_coefficient  consommation_level_1  \
0                 203                    1                    82   
1                 203                    1                  1200   
2                 203                    1                   123   
3                 207                    1                   102   
4                 207                    1                   572   
...               ...                  ...                   ...   
4476744           202                    1                   400   
4476745           202                    1                   200   
4476746           203                    1                   259   
4476747           203                    1                   603   
4476748           203                    1                   516   

         consommation_level_2  consommation_level_3  consommation_level_4  \
0                           0                     0                     0   
1                         184                     0                     0   
2                           0                     0                     0   
3                           0                     0                     0   
4                           0                     0                     0   
...                       ...                   ...                   ...   
4476744                   135                     0                     0   
4476745                     6                     0                     0   
4476746                     0                     0                     0   
4476747                     0                     0                     0   
4476748                     0                     0                     0   

         old_index  new_index  months_number  counter_type  
0            14302      14384              4             0  
1            12294      13678              4             0  
2            14624      14747              4             0  
3            14747      14849              4             0  
4            15066      15638             12             0  
...            ...        ...            ...           ...  
4476744       3197       3732              8             0  
4476745       3732       3938              4             0  
4476746      13884      14143              4             0  
4476747      13281      13884              4             0  
4476748      12765      13281              4             0  

[4476749 rows x 14 columns]